Modules/zlibmodule.c

```diff
@@ -1134,11 +1134,13 @@ zlib_Decompress_flush_impl(compobject *self, Py_ssize_t length)
         return NULL;
     }
 
-    if (PyObject_GetBuffer(self->unconsumed_tail, &data, PyBUF_SIMPLE) == -1)
-        return NULL;
-
     ENTER_ZLIB(self);
 
+    if (PyObject_GetBuffer(self->unconsumed_tail, &data, PyBUF_SIMPLE) == -1) {
```

Comment:
Sorry, I doubt if the patch is correct.

`ENTER_ZLIB()` calls `Py_BEGIN_ALLOW_THREADS` which releases GIL.
In turn, any Python API call without GIL is prohibited, the only available operation is getting GIL back.
Am I correct?

Reply:
`ENTER_ZLIB()` acquires the thread lock, it releases GIL only for `PyThread_acquire_lock((obj)->lock, 1);` invoke, and take GIL back immediately.

```c
#define ENTER_ZLIB(obj) \
    Py_BEGIN_ALLOW_THREADS; \
    PyThread_acquire_lock((obj)->lock, 1); \
    Py_END_ALLOW_THREADS;
#define LEAVE_ZLIB(obj) PyThread_release_lock((obj)->lock);
```

The zlib decompression is performed here: 
```c
            Py_BEGIN_ALLOW_THREADS
            err = inflate(&self->zst, flush);
            Py_END_ALLOW_THREADS
```

Reply:
In 3.10+, `ENTER_ZLIB()` doesn't release GIL if the lock can be acquired immediately:
```c
#define ENTER_ZLIB(obj) do {                      \
    if (!PyThread_acquire_lock((obj)->lock, 0)) { \
        Py_BEGIN_ALLOW_THREADS                    \
        PyThread_acquire_lock((obj)->lock, 1);    \
        Py_END_ALLOW_THREADS                      \
    } } while (0)
#define LEAVE_ZLIB(obj) PyThread_release_lock((obj)->lock);
```